### PR TITLE
Create AttributeSet for use by Nodes and Ports

### LINKED ISSUE #81: Create a consistent abstraction for (key-value) attributes

### PR DESCRIPTION
**Pull request by TysonAndre**
Not doing verification against list of possible attribute names and
types yet.

Fixes  #81

### CODE REVIEW

**Comment by mtrberzi** [comment]
I think this is a good start. This will be a useful class to have when attribute type verification does become a thing, which now becomes trivial because this class could just take a `Map<String, Type>` in order to know what attributes are allowed and what type each value should have.

**Comment by lucaswoj** [comment]
Right now, this is a simple wrapper around a `Map` -- it'll be much more interesting when combined with #27 and #80. I'm ok merging this, pending a few quick changes requested in the comments. Lets start on those two tasks next!

File: src/main/java/org/whdl/intermediate/AttributeSet.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Map;
+
+public class AttributeSet {
+  private Map<String, Value> data;
+
+  public AttributeSet() {
+    data = new HashMap<String, Value>();
+  }
+
+  public Value get(String attrName) throws UndeclaredIdentifierException {
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/AttributeSet.java]: We might want a new exception type for this case. Thoughts?

File: src/main/java/org/whdl/intermediate/AttributeSet.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    if (data.containsKey(attrName)) {
+      return data.get(attrName);
+    } else {
+      throw new UndeclaredIdentifierException("no attribute named '" + attrName
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/AttributeSet.java]: We should not error pass strings directly into exceptions like this. Not your fault, Tyson, I'll make new issues to address the problem.

File: src/main/java/org/whdl/intermediate/AttributeSet.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package org.whdl.intermediate;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class AttributeSet {
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/AttributeSet.java]: How about just calling this class `Attributes`? Its not _exactly_ a set.

File: src/main/java/org/whdl/intermediate/AttributeSet.java (hunk 2 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/AttributeSet.java]: Instantiating `UndeclaredIdentifierException` takes the signature `public UndeclaredIdentifierException(String identifier)` so, in theory, you should just be passing `attrName`. I still think we should create a new exception type for this case but I don't have  a strong opinion.

**Comment by mtrberzi** [comment]
Can this also be implemented on `Connection`s and `Constraint`s? It would be better for the sake of completeness to deal with all of these things at the same time.

**Comment by TysonAndre** [comment]
Made the suggested fixes.
- renamed AttributeSet to Attributes
- added UndeclaredAttributeException
- changed Constraint/Connection to use Attributes object

Probably should have merged against master while reviewing, then rebased before committing instead of rebasing early on.

**Comment by mtrberzi** [comment]
Looks good to me. Merge!

**Pull request merged**